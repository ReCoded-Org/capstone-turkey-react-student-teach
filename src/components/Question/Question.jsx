import { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { FaEllipsisV, FaImage } from 'react-icons/fa';
import questionPhoto from '../../assets/images/questionImage.png';
import personPhoto from '../../assets/images/avatar.jpg';

export const data = [
  {
    id: 1,
    avatar: personPhoto,
    questionImage: questionPhoto,
    name: 'Maria1223',
    questionTitle: 'Lorem Ipsum is simply dummy ',
    questionText:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys 1500s test clamp data classname Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys 1500s test clamp data classname Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys 1500s test clamp data classname',
  },
];

function Question({
  avatar = `${data[0].avatar}`,
  questionImage = data[0].questionImage,
  userName = data[0].name,
  questionTitle = data[0].questionTitle,
  questionText = data[0].questionText,
}) {
  const [open, setOpen] = useState(false);
  const sideMenuRef = useRef();
  const ellipsisIconRef = useRef();
  const [fullImageSize, setFullImageSize] = useState(false);

  useEffect(() => {
    const checkIfClickedOutsideOfSideMenu = (e) => {
      if (
        open &&
        !sideMenuRef.current.contains(e.target) &&
        !ellipsisIconRef.current.contains(e.target)
      ) {
        setOpen(false);
      }
    };
    document.addEventListener('mousedown', checkIfClickedOutsideOfSideMenu);
    return () => {
      document.removeEventListener(
        'mousedown',
        checkIfClickedOutsideOfSideMenu,
      );
    };
  }, [open]);

  return (
    <div className=" relative max-w-3xl mx-auto">
      <div className=" mx-auto my-8 max-w-3xl px-6 pt-4 pb-6 bg-[#F0F0F0]">
        <div className="flex mb-5">
          <img src={avatar} className="w-10 h-10 rounded-full" alt="user.png" />
          <p className="pl-2 font-semibold self-center text-sm mr-auto">
            {userName}
          </p>
          <div
            ref={ellipsisIconRef}
            className="text-xl   self-center cursor-pointer"
          >
            <FaEllipsisV onClick={() => setOpen(!open)} />
          </div>

          <div
            ref={sideMenuRef}
            className={`absolute -right-0 top-12 lg:-right-48 lg:top-0 z-10 w-44 rounded-lg shadow-lg  border-[#CA7560] border-[4px]  border-opacity-50 bg-white transition  py-1 lg:py-3 ${
              open ? 'show opacity-100' : 'hidden opacity-0'
            }`}
            id="dropdown"
          >
            <button
              className=" w-full lg:w-full text-sm text-left text-gray-800 hover:text-gray-900 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
              type="button"
            >
              Edit Question
            </button>
            <button
              className="w-full lg:w-full text-sm text-left hover:bg-gray-200 transition px-2 lg:px-4 py-1"
              type="button"
            >
              Add to Bookmark
            </button>
            <button
              className="w-full lg:w-full text-sm text-left  hover:bg-gray-200 transition px-2 lg:px-4 py-1"
              type="button"
            >
              Share
            </button>
            <button
              className="w-full lg:w-full text-sm text-left text-red-600 hover:text-red-700 hover:bg-gray-200 transition px-2 lg:px-4 py-1"
              type="button"
            >
              Delete Question
            </button>
          </div>
        </div>

        <h3 className="text-2xl mb-5">
          <span className="text-[#CA7560] ">Q</span>: {questionTitle}
        </h3>
        <p className="text-sm mb-7 pl-7 line-clamp-3">{questionText}</p>

        <div
          className={
            fullImageSize
              ? 'fixed top-0 bottom-0 left-0 right-2 lg:left-80 lg:right-0 w-screen h-screen flex items-center z-50'
              : 'hover:scale-105 transition-transform mx-auto w-1/2'
          }
        >
          <button
            onClick={() => setFullImageSize(!fullImageSize)}
            type="button"
          >
            <img
              className=" rounded-xl transition-transform hover:scale-105 hover:cursor-pointer z-0 "
              src={questionImage}
              alt="question.png"
            />
          </button>
        </div>
      </div>

      <div className="relative w-full px-2 lg:px-20 ">
        <input
          className=" line-clamp-5 pr-14 md:pr-16 lg:pr-14 focus:outline-none border-[#CA7560] border-[3px] border-opacity-50 rounded-lg py-4 pl-5 text-xs w-full "
          type="text"
          placeholder="Answer the question..."
        />
        <FaImage className="absolute top-[19px] text-gray-600 cursor-pointer text-lg right-8 md:right-12 lg:right-28 hover:scale-105" />
      </div>
    </div>
  );
}
Question.propTypes = {
  avatar: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  questionTitle: PropTypes.string.isRequired,
  questionText: PropTypes.string.isRequired,
  questionImage: PropTypes.string.isRequired,
};

export default Question;
