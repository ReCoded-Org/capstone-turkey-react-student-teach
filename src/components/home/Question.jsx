import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { useEffect, useState } from 'react';

function Question({ id, profileImage, question, answer, student }) {
  const darkMode = useSelector((state) => state.darkModeReducer.darkMode);
  const [user, setUser] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(
          `https://studentsteach.re-coded.com/api/tutors/${student}`,
        );
        const json = await response.json();
        setUser(json);
      } catch (error) {
        console.log('error', error);
      }
    };
    fetchData();
  }, []);
  console.log(user);
  return (
    <div className="flex flex-col lg:even:float-right mb-4 xl:mb-12 w-12/12 lg:w-10/12 h-20 sm:h-28 truncate">
      <div
        className={`group relative  rounded-lg px-5 py-4 ${
          darkMode ? 'bg-secondaryDark' : 'bg-gray-100'
        } `}
      >
        <div className="flex justify-between items-center mb-2 ">
          <div className="flex items-center  ">
            <a href="/user-profile">
              <img
                className=" w-8 h-8 sm:w-10 xl:w-16 sm:h-10 xl:h-16 rounded-full mr-8 sm:mr-10 xl:mr-20"
                src={user?.avatar === undefined ? profileImage : user.avatar}
                alt="s profile pic"
              />
            </a>
            <a href={`/question/${id}`}>
              <div className="text-sm sm:text-base xl:text-xl  truncate w-[15rem] sm:w-[30rem] lg:w-[40rem] xl:w-[50rem] 2xl:w-[60rem]">
                <span className="text-orange">Q</span>: {question}
              </div>
              <div className="text-sm sm:text-base xl:text-xl text-[#7c7b7b] truncate w-[15rem] sm:w-[30rem] lg:w-[40rem] xl:w-[50rem] 2xl:w-[60rem] ">
                <span className="text-orange ml-4">A</span>: {answer}
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Question;
Question.propTypes = {
  profileImage: PropTypes.string,
  question: PropTypes.string,
  answer: PropTypes.string,
  student: PropTypes.string,
  id: PropTypes.string,
};
Question.defaultProps = {
  profileImage: '',
  question: '',
  answer: '',
  student: '',
  id: '',
};
