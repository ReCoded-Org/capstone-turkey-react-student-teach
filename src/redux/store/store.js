import { configureStore } from '@reduxjs/toolkit';
import addQuestionSlice from '../features/addQuestionSlice';

const store = configureStore({
  reducer: {
    addQuestionReducer: addQuestionSlice,
    signIn: userSlice,
  },
});

export default store;
