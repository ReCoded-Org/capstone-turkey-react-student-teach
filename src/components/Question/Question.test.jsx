import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Question from './Question';
import store from '../../redux/store/store';

it('renders an Question component correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Question
          questionTitle="Loeerasdsa"
          userName="hazarcan"
          avatar="https://via.placeholder.com/150"
          questionText="questionText"
          questionImage="https://via.placeholder.co"
        />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
