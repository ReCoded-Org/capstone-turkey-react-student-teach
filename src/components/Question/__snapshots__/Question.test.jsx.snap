// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an Question component correctly 1`] = `
<div
  className="relative max-w-3xl mx-auto"
>
  <div
    className="mx-auto my-8 max-w-3xl px-6 pt-4 pb-6 rounded-md bg-[#F0F0F0]  text-black"
  >
    <div
      className="flex  justify-between mb-5"
    >
      <div
        className="flex whitespace-nowrap"
      >
        <img
          alt="user.png"
          className="w-10 h-10 rounded-full"
          src="https://via.placeholder.com/150"
        />
        <div
          className="mr-[5rem] lg:mr-0 self-center lg:flex lg:flex-row-reverse"
        >
          <p
            className="text-green-500 ml-4 text-xs  self-center  hidden "
          >
            Author
          </p>
          <p
            className="pl-2 font-semibold self-center text-sm mr-auto"
          >
            hazarcan
          </p>
        </div>
      </div>
      <div
        className="text-xl flex mb-[1px] self-center cursor-pointer whitespace-nowrap"
      >
        <p
          className="mr-3 text-xs  self-center cursor-text"
        >
          
        </p>
        <svg
          className="hidden "
          fill="currentColor"
          height="1em"
          onClick={[Function]}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 192 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
          />
        </svg>
      </div>
      <div
        className="absolute -right-0 top-12 lg:-right-48 lg:top-0 z-10 w-44 rounded-lg shadow-lg transition-all ease-in-out  py-1 lg:py-3 border-[1px] lg:border-0 hidden opacity-0  bg-[#F0F0F0]"
        id="dropdown"
      >
        <button
          className="w-full lg:w-full text-sm text-left text-red-600 hover:text-red-700 hover:bg-gray-200 transition px-2 lg:px-4 py-1 "
          onClick={[Function]}
          type="button"
        >
          Delete Question
        </button>
      </div>
    </div>
    <a
      href="/"
      onClick={[Function]}
    >
      <h3
        className="text-2xl mb-5"
      >
        <span
          className="text-[#CA7560] "
        >
          Q
        </span>
        : 
        Loeerasdsa
      </h3>
      <p
        className="text-sm mb-7 pl-7 line-clamp-3"
      >
        questionText
      </p>
    </a>
  </div>
  <div
    className="relative w-full px-2 lg:px-20  "
  >
    <input
      className="line-clamp-5 pr-14 md:pr-16 lg:pr-14 focus:outline-none border-[#CA7560] border-[3px] border-opacity-50 rounded-lg py-4 pl-5 text-xs w-full bg-[#F0F0F0] placeholder-slate-400"
      placeholder="Answer the question..."
      type="text"
    />
  </div>
</div>
`;
